0000-error-values-proposal.rst

```diff
@@ -0,0 +1,282 @@
+            }
+
+    type MsgDoc = SDoc
+
+    -- for completeness: warnings are represented with the same type
+    -- as errors.
+    type WarnMsg = ErrMsg
+
+where ``SDoc`` is a type from GHC's pretty printing infrastructure that
+represents configurable textual documents.
+
+GHC then maintains a bag of ``ErrMsg`` and a bag of ``WarnMsg`` as
+compilation proceeds and reports them when appropriate.::
+
+    data TcLclEnv = TcLclEnv
+      { ...
+      , tcl_errs :: TcRef Messages     -- Place to accumulate errors
+      , ...
+      }
+
+    type Messages        = (WarningMessages, ErrorMessages)
+    type WarningMessages = Bag WarnMsg
+    type ErrorMessages   = Bag ErrMsg
+
+We propose to replace ``ErrDoc`` with several algebraic data types, each
+representing the different errors/warnings that might arise from a given
+GHC subsystem. For example (simplified):::
+
+    data RenamerError
+        = NotInScope OccName [Name] -- unknown name, suggestions
+	| ...
+
+    data TypecheckerError
+        = OccursCheck Type Type
+	| ...
+
+    ...
+
+We could even split error types further if necessary, making it a
+slightly more elaborate/deep hierarchy. The exact shape of the said hierarchy
+has yet to be determined, as it will be best informed by staring at the
+error generation code that GHC has today. We would also provide a toplevel
+error data type, ``GHCError``, which would be a sum of the error types from all
+subsystems. This would allow us to store and more generally treat
+uniformly errors from different systems:::
+
+    data GHCError
+      = PsError ParseError
+      | RnError RenamerError
+      | TcError TypecheckerError
+      | ...
+
+One can then define helper functions such as
+``notInScope :: OccName -> Name -> GHCError`` to be able to easily construct
+error values from within, say, the guts of a typechecker function, without
+having to make the intermediate layers and wrapping visible there. We would
+create, manipulate and store ``GHCError`` values until the very last moment,
+when it is time to render the errors and report them. This would require
+implementing ``errorMessage :: GHCError -> ErrMsg``, and would be equivalent
+to all the ``ErrMsg`` building code that GHC has right now.
+
+Error consumers (the GHC program, GHC API users) would be presented with
+``GHCError`` values, and would be free to just call ``errorMessage`` on them
+to generate error message documents, or do something more interesting with
+some or all errors, using good old pattern matching to provide a specific
+interpretation for the errors of interest.
+
+For error producers, the main change is that the different subsystems will
+define error types and helper functions to build error values. In order for
+``GHCError`` to be able to refer to all the different error types, and for
+those types to use names from the module where ``GHCError`` is defined, we will
+have to introduce a few ``.hs-boot`` files to work around the import cycles
```

Comment:
Expect push-back from Simon on this point. So it might be worth sketching out exactly what hs-boot files we'll need.

I think we'll need precisely 1, declaring `data GHCError`. Each sub-system has a module that defines a bunch of its internal types; these modules will also now define e.g. `TcError` or `PsError`. I don't think hs-boot files will be necessary -- unless the sub-system already has a strange topography. Each sub-system will also define e.g. `renderTcError :: TcError -> SDoc` (I think it won't be quite so simple, but you get the idea). Then, a new module `Error`, in the `main` directory, defines `GHCError` once and for all.

This plan does sorely interfere with the goal of splitting GHC up into packages. But I think either Backpack or my "abstract data" idea can fix that problem when it starts to hurt.

Does this sound like it will work to you? As I write this, I worry that I'm overly optimistic about how easy this aspect will be, so please do have a think.

Reply:
Well, let's spell things out. At a very high level, ignoring imports, we have:

``` hs
-- compiler/typecheck/TcError.hs (or TcErrors.hs, which already exists)
data TcError = ...
renderTcError :: TcError -> SDoc
```

``` hs
-- compiler/main/Error.hs
data GHCError = TcErr TcError | ...
renderGHCError :: GHCError -> SDoc
```

So far, no cycle. We can just `import TcError` in `Error` to bring the `TcError` type (for the `TcErr` constructor) and `renderTcError` (for the `TcErr` clause in the definition of `renderGHCError`) in scope.

We presumably want to store bags of `GHCError`/`GHCWarning` values instead of the current `Messages` type, in `tcl_errs`. It could have been typechecker errors only, but the same monad is at the very least used for renaming and typechecking IIRC. Which means that at some point, the typechecker code needs to turn a `TcError` into a `GHCError`. Module where such code would live would need to both construct a `TcError` (hence needs the `TcError` module) and turn it into a `GHCError` (hence the need to import `Error`) using the `TcErr` constructor. `Error` in turn imports `TcError`, but that seems okay, no cycle...? I don't think we actually need to expose anything related to `GHCError` to the respective subsystems. In case we _do_ want to supply some helpers for constructing typechecker errors as `GHCError` values "directly", then that's the point where a cycle might show up, if we put such helpers in `Error` or `TcError`: if the former, we'd need access to `TcError` constructors, we already import that module, it's fine; if the latter, we'd need to bring `GHCError` and its constructors in scope => cycle.

So it looks like there are ways to work around the cycles without hs-boot files...? I'm surely missing something.

Reply:
We'll need to mention `GHCError` in the environment, right? That's the cycle I was worried about. But you're right that we'll also need to have a function somewhere that converts a `TcError` to a `GHCError`. That will introduce more cycles. :(

Maybe the type-checker env't really stores `TcError`s, with some higher-level module converting them to `GHCError`s.

Or maybe the high-level `Error.hs` defines `GHCError`, `TcError`, and all the rest. That should be possible because it sits near the top of the hierarchy. It also defines conversion functions from e.g. `TcError` to `GHCError`. Then much of this is exported through its hs-boot file. So we'll have one swollen hs-boot file instead of many. I've lost some of my optimism, but I still favor the closed-datatype approach over the open one.

Reply:
Yes, `GHCError` will be used in the local env, in `TcRnTypes`/`TcRnMonad`/`TcRnDriver`.  But those modules won't be imported by `TcError` will they? I'm not entirely sure we can't work around the cycle just yet.

But if we really need at least one, then your very last suggestion doesn't sound like too hard a sell to me.

Reply:
I suppose `TcError` can indeed be rather low down in the hierarchy. This would be good. But `GHCError` will have to be above `TcError` / `PsError` / `RnError` etc. Maybe it can be squeezed in between everything. But then we'll have a disaster trying to separate out components of GHC.

Perhaps this conversation suggests that it's hard to figure this out ahead of time and that we should wait until implementing so that we can have a concrete plan.